### PR TITLE
Core/ts version task 371 create own candles

### PR DESCRIPTION
**Pull request by isaac-fibonacci**
Add own candles system with influxdb

- [ ] All Merge Request pipelines should be passed.
- [ ] Make sure all the changes are necessary and correct.
- [ ] Make sure no needed change missing (all necessary changes are made).
- [ ] Merge Request should contains only changes for one purpose (one feature or bug or refactoring).
- [ ] All necessary unit tests are written.
- [ ] Critical changes should be tested locally (or on staging).

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "[typescript]": {
-    "editor.defaultFormatter": "rvest.vs-code-prettier-eslint"
+    "editor.defaultFormatter": "vscode.typescript-language-features"
```
Comment (hypersolo) [inline .vscode/settings.json:3]: revert this.

@GreatGaryCooper has configured this as our defulat formatter.

File: package-lock.json [hunk 2]
```diff
@@ -9,6 +9,8 @@
       "version": "1.0.0",
       "license": "ISC",
       "dependencies": {
+        "@influxdata/influxdb-client": "^1.33.2",
```
Comment (hypersolo) [inline package-lock.json:12]: changes of these file must be reverted.

it's for js-version not ts-version.

File: ts-version/.env.example [hunk 3]
```diff
@@ -46,6 +46,24 @@ DB_DIALECT='postgres'
 # ---------------------------------------------------------------------
 
 
+
```
Comment (hypersolo) [inline ts-version/.env.example:49]: add these to env.ts and env-validation.ts too.

File: ts-version/src/presentation/http/express.ts [hunk 4]
```diff
@@ -89,6 +91,7 @@ export default class HttpServer {
     this.expressApp.delete("/user/orders", auth, (req, res, next) =>
       this.orderController.cancelAll(req, res).catch(next),
     );
+    this.expressApp.get('candles', (req, res) => this.candlesController.get(req, res))
```
Comment (hypersolo) [inline ts-version/src/presentation/http/express.ts:94]: handle async error

File: ts-version/src/presentation/http/controllers/candle.controller.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+export class CandlesController extends BaseController {
+    constructor(private service: IDataService.schema) {
+        super();
+    }
+
+    public async get(req: Request, res: Response) {
+        const { network, market, start, end, duration = "24h" } = req.query;
+
+        const measurement = {
```
Comment (hypersolo) [inline ts-version/src/presentation/http/controllers/candle.controller.ts:14]: what is the purpose of this?

File: ts-version/src/domain/ports/data-service.interface.ts [hunk 6]
```diff
@@ -7,6 +8,7 @@ import Order from "@entity/order/order.entity";
 import { FillSide } from "@type/fill.type";
 import { Nullable } from "@type/nullable.type";
 import { SupportedNetworks } from "@type/supportedNetwrok";
+import { type } from "os";
```
Comment (hypersolo) [inline ts-version/src/domain/ports/data-service.interface.ts:11]: delete this.

File: ts-version/src/domain/entities/candle.entity.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { NumberEngine } from "@util/config/math-engine";
+
+export default class Candle {
```
Comment (hypersolo) [inline ts-version/src/domain/entities/candle.entity.ts:3]: it doesn't have any constructor. in this way when we want a Candle, we must set all fields externally after calling constructor.

if we don't have any functionalities in this class, i think using dto might help.

File: ts-version/src/data/database/influxdb/repository/candle.adapter.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        const point2 = new Point(data.measurement)
+            .floatField('value', data.price)
+            .floatField("volume", data.volume)
+            .booleanField("takerbuy", data.takerbuy)
+            .timestamp(data.date)
+        writeApi.writePoint(point2)
+        try {
+            await writeApi.close()
+        } catch (e) {
+            this.logger.error(e)
```
Comment (hypersolo) [inline ts-version/src/data/database/influxdb/repository/candle.adapter.ts:171]: use error system

File: ts-version/src/data/database/influxdb/repository/candle.adapter.ts [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                candles.push(series as unknown as Candle)
+            }
+        } catch {
+            throw new Error("candle not found")
+        }
+        return candles
+    }
+
+    async write(data: StoreFluxCandleDTO): Promise<void> {
+        const point2 = new Point(data.measurement)
```
Comment (hypersolo) [inline ts-version/src/data/database/influxdb/repository/candle.adapter.ts:162]: point2 >>> point

File: ts-version/src/data/database/influxdb/listeners/candle.listener.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import { IEventObserver } from "@port/event-observer.interface";
+import CandleRepoAdapter from "../repository/candle.adapter";
+import Fill from "@entity/fill.entity";
+
+export default class CandleCreateListener implements IEventObserver {
+    private static candleCreateListener: CandleCreateListener;
+    private static eventName = "fill_created";
+
+
+    update(_newObject: Observable, _oldObject?: Observable) {
```
Comment (hypersolo) [inline ts-version/src/data/database/influxdb/listeners/candle.listener.ts:11]: idea:
using Fill as input type prevent us from type casting in next line

File: .vscode/settings.json (hunk 1 above)
Comment (isaac-fibonacci) [reply .vscode/settings.json:3]: Done

File: ts-version/src/domain/ports/data-service.interface.ts (hunk 6 above)
Comment (isaac-fibonacci) [reply ts-version/src/domain/ports/data-service.interface.ts:11]: Done

File: ts-version/src/presentation/http/controllers/candle.controller.ts (hunk 5 above)
Comment (isaac-fibonacci) [reply ts-version/src/presentation/http/controllers/candle.controller.ts:14]: I don't understand what you mean

File: .vscode/settings.json (hunk 1 above)
Comment (isaac-fibonacci) [reply .vscode/settings.json:3]: Done

File: ts-version/src/presentation/http/express.ts (hunk 4 above)
Comment (isaac-fibonacci) [reply ts-version/src/presentation/http/express.ts:94]: Done

File: ts-version/src/data/database/influxdb/repository/candle.adapter.ts (hunk 9 above)
Comment (isaac-fibonacci) [reply ts-version/src/data/database/influxdb/repository/candle.adapter.ts:162]: Done

File: package.json [hunk 11]
```diff
@@ -15,6 +15,8 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
+    "@influxdata/influxdb-client": "^1.33.2",
```
[the remark this answers is not in the document]
Comment (isaac-fibonacci) [reply package.json:18]: Done

File: package-lock.json (hunk 2 above)
Comment (isaac-fibonacci) [reply package-lock.json:12]: Done

File: ts-version/src/domain/entities/candle.entity.ts (hunk 7 above)
Comment (isaac-fibonacci) [reply ts-version/src/domain/entities/candle.entity.ts:3]: DTO responsibility is Data Transfer Object, but the candle in our code is a stand-alone data entity

File: ts-version/.env.example (hunk 3 above)
Comment (isaac-fibonacci) [reply ts-version/.env.example:49]: Done

File: ts-version/src/data/database/influxdb/influxdb.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { InfluxDB } from "@influxdata/influxdb-client";
+const options = {
+    url: `http://${process.env.INFLUXDB_HOST}:${process.env.INFLUXDB_PORT}`,
```
[the remark this answers is not in the document]
Comment (isaac-fibonacci) [reply ts-version/src/data/database/influxdb/influxdb.ts:3]: Done

File: ts-version/src/data/database/influxdb/repository/candle.adapter.ts [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+    async read(params: ICandleRepo.GetFluxCandleDTO): Promise<Candle[]> {
+        let candles: Candle[] = []
+        try {
+            for await (const { values, tableMeta } of queryApi.iterateRows(this.getFluxQuery(params))) {
+                const series = tableMeta.toObject(values)
+                candles.push(series as unknown as Candle)
+            }
+        } catch {
+            throw new Error("candle not found")
```
[the remark this answers is not in the document]
Comment (isaac-fibonacci) [reply ts-version/src/data/database/influxdb/repository/candle.adapter.ts:156]: Done

File: ts-version/src/data/database/influxdb/listeners/candle.listener.ts (hunk 10 above)
Comment (isaac-fibonacci) [reply ts-version/src/data/database/influxdb/listeners/candle.listener.ts:11]: Done

File: ts-version/src/presentation/http/controllers/candle.controller.ts (hunk 5 above)
Comment (isaac-fibonacci) [reply ts-version/src/presentation/http/controllers/candle.controller.ts:14]: Done

**Review by hypersolo — APPROVED** [review]

**Pull request merged**